protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+The specification of consensus implicitly considers `valid(v)` as an immutable function,
+which is always supposed to provide the same result, when called at the same height
+for the same value `v`.
+This was the main reason for introducing the determinism requirements on `ProcessProposal`
+in the ABCI 1.0 specification.
+
+If we are to relax the determinism requirements on the application,
+we first need to weaken the `valid(v)` predicate and therefore consensus _validity_.
+In the proposed solution, correct processes MAY evaluate `valid(v)` to _true_ or _false_ at different times
+for the same value `v` during height `h`.
+
+Consensus _validity_ invariant is then modified as follows:
+
+- _weak validity_: predicate `valid(v)` has been evaluated to _true_ at least once by one correct process
+  for the decided block `v`.
+
+### Eventual requirements
+
+We now relax the relevant ABCI 1.0 requirements in the following way.
+
+* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:
+  There exists a round $r_s \ge 0$ of height $h$ such that,
+  for any two correct processes $p$ and $q$ and any round $r_p \geq r_s$,
+  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
+  $q$'s application returns Accept in `ResponseProcessProposal`.
+
+* The determinism-related requirements, namely requirements 4 and 5, are removed.
+
+We call round $r_s$ the coherence-stabilization round.
+
+> If we think in terms of `valid(v, bc(h-1), x_p)`, observe that it is in the responsibility of the application to
+> ensure 3b, that is, the application designers need to prove that the `x_p` values at correct processes 
+> are over time evolving
+> in a way that eventually ResponseProcessProposal returns ACCEPT.
+> In analogy to proposer based time, `x_p` can be considered to be process `p`s local clock, and having
+> clocks synchronized was used to prove acceptance of a proposal.
+
+> [TODO: can we _tighten up_ eventual coherence?... i.e., make it even weaker?]
+>
+> [TODO: Since we have removed determinism, can byzantine proposers cause mayhem now?]
+
+### Modifications to the consensus Algorithm
+
+The Tendermint algorithm as described in the [arXiv paper][arxiv],
+and as implemented in CometBFT up to version `v0.38.0-rc3`,
+cannot guarantee consensus _termination_ for applications
+that just fulfill requirement 3b (eventual coherence), but do not fulfill requirements 3, 4, and 5 .
+
+We need the following modifications (in terms of the algorithm as described in page 6 of the arXiv paper):
+
+- remove the evaluation of `valid(v)` in lines 29, 36 and 50 (i.e. replace `valid(v)` by `true`)
+- modify line 23 as follows
+
+> _\[Original\]_ &nbsp; 23: **if** $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$ **then**
+
+> _\[Modified\]_
+>
+> &nbsp; 23a: $validValMatch := (validRound_p \neq -1 \land validValue_p = v)$
+>
+> &nbsp; 23b: **if** $[lockedRound_p = −1 \land (validValMatch \lor valid(v))] \lor lockedValue_p=v$ **then**
```

Comment:
Can you elaborate the rationale here?

From my understanding, we accept values that are either locked or valid (this is introduced by the change). But why?

Reply:
The only difference of locked and valid is that we voted (issued a `Precommit`) for the `lockedValue`.

Reply:
These are good points, and make me realise we should explain the "why" in the text. Let me see what I can do

Reply:
The idea is that a node uses the information that a value v was considered valid before (validValue == v) even if the node itself has not voted on it before (lockedRound == -1), thus avoiding invoking the _valid_ function. In PBTS terms, this would avoid considering a good proposal invalid if it is repeated in rounds bigger than the one it was first proposed.

Reply:
I think I understood the formula wrong, but we should not accept a value just because it matches our valid value.

For the part that we might skip the (non-deterministic) validity check in this case, I agree. I still believe that the (deterministic) validity check must be performed at least once for a committed block.